state-synchronizer/src/coordinator.rs

```diff
@@ -361,163 +427,171 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
-                            req.ledger_info_with_sigs = Some(sync_req.target.clone().into());
-                        }
-                        0
-                    }
-                    RoleType::FullNode => {
-                        req.timeout = self.config.long_poll_timeout_ms;
-                        self.config.long_poll_timeout_ms
-                    }
-                };
-                debug!(
-                    "[state sync] request next chunk. peer_id: {}, known_version: {}, timeout: {}",
-                    peer_id,
-                    self.known_version + offset,
-                    timeout
-                );
-
-                let msg = StateSynchronizerMsg {
-                    message: Some(StateSynchronizerMsg_oneof::ChunkRequest(req)),
-                };
+    /// Sends a chunk request with a given `known_version` (might be chosen optimistically).
+    /// The request includes a target for Validator and a non-zero timeout for a FullNode.
+    async fn send_chunk_request(&mut self, known_version: u64) {
+        let (peer_id, mut sender) = match self.peer_manager.pick_peer() {
+            Some((p, s)) => (p, s),
+            None => {
+                return;
+            }
+        };
 
-                if sender.send_to(peer_id, msg).await.is_err() {
-                    error!("[state sync] failed to send p2p message");
+        let mut req = GetChunkRequest::default();
+        req.known_version = known_version;
+        req.limit = self.config.chunk_limit;
+        if self.role == RoleType::Validator {
+            let target = match self.sync_request.as_ref() {
+                Some(req) => req.target.clone().into(),
+                None => {
+                    error!("[state sync] Validator chunk request without a sync request.");
+                    return;
                 }
-                counters::OP_COUNTERS.inc(&format!("{}.{}", counters::REQUESTS_SENT, peer_id));
-            }
+            };
+            req.ledger_info_with_sigs = Some(target);
+        } else {
+            req.timeout = self.config.long_poll_timeout_ms;
         }
-    }
+        debug!(
+            "[state sync] request next chunk. peer_id: {}, known_version: {}, timeout: {}",
+            peer_id.short_str(),
+            known_version,
+            req.timeout
+        );
+        let msg = StateSynchronizerMsg {
+            message: Some(StateSynchronizerMsg_oneof::ChunkRequest(req)),
+        };
 
-    async fn store_transactions(
-        &self,
-        txn_list_with_proof: TransactionListWithProof,
-        ledger_info: LedgerInfoWithSignatures,
-    ) -> Result<()> {
-        self.executor_proxy
-            .execute_chunk(txn_list_with_proof, ledger_info)
-            .await
+        self.peer_manager
+            .process_request(known_version + 1, peer_id);
+
+        if sender.send_to(peer_id, msg).await.is_err() {
+            error!("[state sync] failed to send p2p message");
+        }
```

Comment:
ensure?

Reply:
at the layer we don't want to propagate the error, we need `error` to log instead of `ensure`, I feel the main loop might be a good place

Reply:
actually looking at this, the fact that we have await occurring it is probably more readable than placing this within yet another function / macro.